### DISCUSSION TITLE
bxcan Transmit Interrupt

### DISCUSSION THREAD

**Discussion by nopnop2002** [stm32duino/Arduino_Core_STM32#2120] [Q&A]
I'm using NUCLEO STM32F103 and Version 2.6.0 Core.
I want to use the BxCAN transmission completion interrupt

Four interrupt vectors are dedicated to bxCAN. 
Each interrupt source can be independently enabled or disabled by means of the CAN Interrupt Enable register (CAN_IER).

My Code:

Set the IER register during BxCAN initialization.

```
  CAN1->IER  |=   0x1UL;                // Set to Transmit mailbox empty interrupt enable

```

add an interrupt handler.
```
void CAN1_TX_IRQHandler (void)
{
  Serial.println("USB_HP_CAN1_TX_IRQHandler");  
}
```

Add the following to the Setup function.

```
  NVIC_SetPriority(CAN1_TX_IRQn, 1);
  NVIC_EnableIRQ(CAN1_TX_IRQn);
```

but it doesn't work properly.
Sending BxCAN freezes the sketch.

If I comment out the following, it works fine.
```
  //CAN1->IER  |=   0x1UL;                // Set to Transmit mailbox empty interrupt enable

```
I would like some advice.

**Comment by fpistm** [comment]
Hi @nopnop2002 
I'm not an expert on CAN. 
Anyway, for me it seems you define the `CAN1_TX_IRQHandler` in ino file so you have to add `extern "C"` before else the `Default_Handler` is called (infiinite loop).
Also I do not recommend using `Serial` in an IRQ handler as `Serial` also uses IT.

**Comment by nopnop2002** [comment]
**Thanks for your comment.**

I changed it like this.
But it hangs.
```
extern "C" void CAN1_TX_IRQHandler (void)
{
  //Serial.println("USB_HP_CAN1_TX_IRQHandler");  
}
```



It seems that CAN1_TX_IRQHandler(same as USB_HP_CAN1_TX_IRQHandler)  is defined as a week function.
```
system\Drivers\CMSIS\Device\ST\STM32F1xx\Include\stm32f103x6.h(10087): #define USB_HP_IRQHandler             USB_HP_CAN1_TX_IRQHandler
system\Drivers\CMSIS\Device\ST\STM32F1xx\Include\stm32f103x6.h(10088): #define CAN1_TX_IRQHandler            USB_HP_CAN1_TX_IRQHandler


system\Drivers\CMSIS\Device\ST\STM32F1xx\Source\Templates\gcc\startup_stm32f103x6.s(166):   .word USB_HP_CAN1_TX_IRQHandler
system\Drivers\CMSIS\Device\ST\STM32F1xx\Source\Templates\gcc\startup_stm32f103x6.s(292):   .weak USB_HP_CAN1_TX_IRQHandler
system\Drivers\CMSIS\Device\ST\STM32F1xx\Source\Templates\gcc\startup_stm32f103x6.s(293):   .thumb_set USB_HP_CAN1_TX_IRQHandler,Default_Handler
```


Even if we remove this, no errors will occur in the link.
```
# if 0
extern "C" void CAN1_TX_IRQHandler (void)
{
  //Serial.println("USB_HP_CAN1_TX_IRQHandler");  
}
#endif
```

but it hangs.


The implementation of the week function may be incomplete.

**Reply by fpistm** [reply]
> It seems that CAN1_TX_IRQHandler(same as USB_HP_CAN1_TX_IRQHandler) is defined as a week function.

Right, all IRQ handler have a weak function which by default call the Default handler.

The right cmsis  device file is stm32f103xb not stm32f103x6. Same for the startup file.

https://github.com/stm32duino/Arduino_Core_STM32/blob/0c69f378ddf18937fdbdb1f26f61e1308ac7ad51/system/Drivers/CMSIS/Device/ST/STM32F1xx/Include/stm32f103xb.h#L10199C1-L10200C1

Just to avoid any issue, you do not use USB?

For the redefinition of the Weak, it should works as expected. You can check in the elf file. But I have no doubt it is correct. You used Arduino IDE ? not PIO?

**Comment by nopnop2002** [comment]
>Just to avoid any issue, you do not use USB?

I use NUCLEO-STM32F103.
So i use usb for firmware flash and Serial.print.
There is no code in my code that manipulates USB registers like USB_CNTR/USB_ISTR.


>You used Arduino IDE ? not PIO?

I use Arduino IDE.

![Arduino-IDE](https://github.com/stm32duino/Arduino_Core_STM32/assets/6020549/ea49db77-1d52-4d48-b65c-09df01dd7dff)

![stm32-core](https://github.com/stm32duino/Arduino_Core_STM32/assets/6020549/6846b949-5859-4a12-8fb2-3d277d5c06d4)

**Comment by fpistm** [comment]
IIWR you can't use USB and CAN at the same time.

Or when you told Serial uses USB you talk about the one from STLink

**Comment by nopnop2002** [comment]
Serial objects can be used as standard in NUCLEO.
In NUCLEO Serial objects are tied to USB-Port.

>IIWR you can't use USB and CAN at the same time.

Are you saying that BxCAN and Serial objects cannot be used at the same time on the NUCLEO board?

STM32F103 can use a Virtual COM port.
If we enable Virtual COM port, we can use the USB port as a serial port.
But Basic Extended CAN(BxCan) and Virtual COM port cannot be used at the same time.

In NUCLEO, i think that it not use Virtual COM port.
I think Virtual COM port and STLINK COM port are different.


This is the Arduino Serial monitor when performing BxCAN communication using the NUCLEO board.
Serial.print gose to NUCLEO USB port.
When not using interrupts, all BxCAN communications work normally.

![Serial print](https://github.com/stm32duino/Arduino_Core_STM32/assets/6020549/c2997f1a-9e20-4fa3-8666-b457ce0803c2)

Are you saying that [BxCAN Inetrrupt function] and [STLINK COM port Serial objects] cannot be used at the same time on the NUCLEO board?

**Reply by fpistm** [reply]
No, my point was if you used SerialUSB but you used an USART which is connected to the embedded STLink and forward to VCP of the Nucleo so no issue here.

**Reply by nopnop2002** [reply]
If so, why does it hang with the week function?

**Reply by fpistm** [reply]
the weak function is an infinite loop so it is normal it hangs. But as stated before you should be able to redefine but don't know why it does not work for you. Probably another issue but don't know which one. Your best best is to debug. With Arduino IDE2 you can debug easily.

**Reply by fpistm** [reply]
I've tried build with and without CAN1_TX_IRQHandler defined:

Without:
0800233c W USB_HP_CAN1_TX_IRQHandler
With:
0800015c T USB_HP_CAN1_TX_IRQHandler

As you can see using `arm-none-eabi-nm sketch_sep6a.ino.elf `, the result is correct. 
See https://manpages.debian.org/testing/binutils-arm-none-eabi/arm-none-eabi-nm.1.en.html for signification of W and T.

**Reply by fpistm** [reply]
Are you sure the build with extern "C" added was done and do not use cache build?

**Reply by nopnop2002** [reply]
Switch board to Nucleo F401RE

Only Compile HelloWorld.ino

Switch board to Nucleo F103RB

I think we can now compile without caching.

add this code to bxcan test code.
```
extern "C" void CAN1_TX_IRQHandler (void)
{
  //Serial.println("USB_HP_CAN1_TX_IRQHandler");  
}
```

Flash firmware to Nucleo F103RB


Hang.

**Reply by fpistm** [reply]
OK. So you should launch a debug session.

**Comment by nopnop2002** [comment]
I installed Arduino IDE 2.2.1.


![IDE2 2 1](https://github.com/stm32duino/Arduino_Core_STM32/assets/6020549/25e1477e-a5ca-450f-b3ea-f5569099b7f2)

I have never used debug session.

I need advice on what to do.


However, Arduino IDE 2 has a lot of bugs.

**Comment by nopnop2002** [comment]
I set a Breakpoint on CAN1_TX_IRQHandler.

I found out that HALT is done on return from this function.

Stop at break point
![Debuuger-0](https://github.com/stm32duino/Arduino_Core_STM32/assets/6020549/0518404a-1735-490c-8498-52521b82a8a1)

Press StepOver
![Debuuger-1](https://github.com/stm32duino/Arduino_Core_STM32/assets/6020549/9d4263b4-b0fc-4049-be21-91b100dbd930)

**Comment by nopnop2002** [comment]
Has there been any progress?

Or do I need to issue new issues as a core library bug?

**Reply by fpistm** [reply]
Do you have enabled debug symbol and set optimization to -Og?
I'm surprised you are able to break in the handler as you do nothing and I guess compiler should simply ignore the int i declaration as it is not used. In a general way in an handler we used a volatile like this no optimization occured.
I have no more clue for you.  No need to open an issue as we have no CAN library.
You can try to get some input on the forum and also check the errata sheet.
I would also try to convert a STM32CubeF1 CAN examples to see if it works.

**Reply by nopnop2002** [reply]
compile otions is here.


![compile_options](https://github.com/stm32duino/Arduino_Core_STM32/assets/6020549/09ee8b52-1d0b-4522-a692-4bd319289836)

One possibility is that there is a problem with the code that jumps from C++ to a C function for handling interrupts.

>I would also try to convert a STM32CubeF1 CAN examples to see if it works.

I have never used STM32CubeF1, but I think it needs to be built with C++.

**Reply by fpistm** [reply]
Select the options stated before this will help to debug.
